src/discovery/Discovery.ts

```diff
@@ -175,299 +208,291 @@ class Discovery {
+          // If we don't have identity info, simply skip this vertex
+          if (vertexIdentityInfo == null) {
+            return;
+          }
+          // Link the identity with each node from its claims on the provider
+          // Iterate over each of the claims
+          for (const id in vertexIdentityInfo.claims) {
+            if (ctx.signal.aborted) throw ctx.signal.reason;
+            const identityClaimId = id as IdentityClaimId;
+            const claim = vertexIdentityInfo.claims[identityClaimId];
+            // Claims on an identity provider will always be node -> identity
+            // So just cast payload data as such
+            const data = claim.payload.data as ClaimLinkIdentity;
+            const linkedVertexNodeId = nodesUtils.decodeNodeId(data.node)!;
+            const linkedVertexGK =
+              gestaltsUtils.keyFromNode(linkedVertexNodeId);
+            // Get the chain data of this claimed node (so that we can link in GG)
+            let linkedVertexChainData: ChainData;
+            try {
+              linkedVertexChainData = await this.nodeManager.requestChainData(
+                linkedVertexNodeId,
+              );
+            } catch (e) {
+              if (
+                e instanceof nodesErrors.ErrorNodeConnectionDestroyed ||
+                e instanceof nodesErrors.ErrorNodeConnectionTimeout ||
+                e instanceof networkErrors.ErrorConnectionNotRunning
+              ) {
                 if (!this.visitedVertices.has(linkedVertexGK)) {
-                  await this.pushKeyToDiscoveryQueue(linkedVertexGK);
+                  await this.scheduleDiscoveryForVertex(linkedVertexGK);
                 }
+                this.logger.error(
+                  `Failed to discover ${data.node} - ${e.toString()}`,
+                );
+                continue;
+              } else {
+                throw e;
               }
             }
-            break;
-          default:
-            never();
+            // With this verified chain, we can link
+            const linkedVertexNodeInfo: NodeInfo = {
+              id: nodesUtils.encodeNodeId(linkedVertexNodeId),
+              chain: linkedVertexChainData,
+            };
+            await this.gestaltGraph.linkNodeAndIdentity(
+              linkedVertexNodeInfo,
+              vertexIdentityInfo,
+            );
+            // Add this vertex to the queue if it is not present
+            if (!this.visitedVertices.has(linkedVertexGK)) {
+              await this.scheduleDiscoveryForVertex(linkedVertexGK);
+            }
+          }
         }
-        this.visitedVertices.add(vertex);
-        await this.removeKeyFromDiscoveryQueue(vertexId);
-      }
+        break;
+      default:
+        never();
     }
+    this.visitedVertices.add(vertex);
   }
 
   /**
-   * Will resolve once the queue has drained
+   * Will resolve once all discovery tasks have finished
    */
   public async waitForDrained(): Promise<void> {
```

Comment:
So this function should be renamed and made a bit more precise. It should be about waiting for all currently scheduled, queued and active discovery tasks to finish. It does not signal cancellation to any of the tasks. Therefore it cannot be integrated into the `stop` method.

This is used for testing, so that way we can wait for discovery side-effects to complete.

Let's call this:

```ts
await waitForDiscoveryTasks();
```

This can then be similar to the other function `hasDiscoveryTasks` which just returns a boolean if there are any scheduled, queued or active tasks in the discovery system.

The function should be kept simple in that it only checks once for all the tasks and waits for them. The looping implementation atm cannot guarantee that no more tasks are scheduled and it could also potentially run forever.

Reply:
I'm having these methods return the number of existing tasks just avoid boolean blindness.

Reply:
Then is this `getTaskCount()`?